### PR TITLE
BinarySearch in 2D array with O(n) Time complexity #6513: an pr under Hacktoberfest

### PR DESCRIPTION
**Pull request by sofezticated-gentleman**
In this pr I am submitting my solution for the aforementioned issue under the festival of Hacktoberfest

### Describe your change:
Over here I am submitting my algo for Binsearch for the aforementioned issue

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by sofezticated-gentleman** [comment]
@dhruvmanila can you please review my pr

File: matrix/matrix_binary_search.py
Comment (cclauss) [reply matrix/matrix_binary_search.py:16]: Make it clear that you want an int, by using floor division (//).

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:21]:
```suggestion
    if r in (lower_bound, upper_bound):
```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:23]: As discussed in CONTRIBUTING.md, please run this code thru psf/black to get rid of all these extra parentheses.  This ain't JavaScript.

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:54]: psf/black will get rid of all these unneeded lines.

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:14]: How about a test that returns False?

**Comment by sofezticated-gentleman** [comment]
@cclauss i have done the required changes can you please help me out. I am facing issues in checks but aint able to understand them. where am i going wrong.

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:18]:
```suggestion
    if array[r] == value:

```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:48]:
```suggestion
        index += 1

```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:26]:
```suggestion


def mat_bin_search(value: int, matrix: list) -> bool:
```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:24]:
```suggestion
    else:
```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:13]:
```suggestion
    >>> binary_search(matrix, 0, len(matrix) - 1, 1)
```
Remove the space at the end of the line or get psf/black to do it for you.  See CONTRIBUTING.md.

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:19]:
```suggestion
```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:25]: You NEED two blank lines before and after a function of psf/black will fail your pull request.  See CONTRIBUTING.md
```suggestion


def mat_bin_search(value: int, matrix: list) -> bool:
```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:37]:
```suggestion
    True
```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:34]: Use ... for line continuation in doctests...
```suggestion
    >>> matrix = [[1, 4, 7, 11, 15],
    ...           [2, 5, 8, 12, 19],
    ...           [3, 6, 9, 16, 22],
    ...           [10, 13, 14, 17, 24],
    ...           [18, 21, 23, 26, 30]]
```

**Comment by sofezticated-gentleman** [comment]
@cclauss can you please review my pr, i done the required changes

**Comment by dhruvmanila** [comment]
Why is there another binary search? We already have them!

**Comment by sofezticated-gentleman** [comment]
@dhruvmanila, a issue was opened a day ago for a 2d binary search in a matrix and so i tried to make a pr, also its my second issue, I think so theres no 2d binary search in the matrix repo....

**Review by dhruvmanila — CHANGES_REQUESTED** [review]
Please link the issue to the PR. Search algorithms do not return a boolean value, they return the position of the element.

In this case, it should return the position of the value in the matrix like `[y, x]` where `y` would be the index of the subarray in which the element exists and `x` would be the index of the element in that subarray.

**Comment by sofezticated-gentleman** [comment]
> Please link the issue to the PR. Search algorithms do not return a boolean value, they return the position of the element.
> 
> In this case, it should return the position of the value in the matrix like `[y, x]` where `y` would be the index of the subarray in which the element exists and `x` would be the index of the element in that subarray.

Oh thank you @dhruvmanila I will make those changes ASAP

**Comment by sofezticated-gentleman** [comment]
@dhruvmanila I have done the requested changes, but then too my build tests are failing, what are build tests and why is my code failing in them , can you please help me out...

**Comment by MaximSmolskiy** [comment]
> @dhruvmanila I have done the requested changes, but then too my build tests are failing, what are build tests and why is my code failing in them , can you please help me out...

@sofezticated-gentleman Your second test for `mat_bin_search` function fails (log is listed below).

```
=================================== FAILURES ===================================
_____________ [doctest] matrix.matrix_binary_search.mat_bin_search _____________
034     >>> matrix = [[1, 4, 7, 11, 15],
035     ...           [2, 5, 8, 12, 19],
036     ...           [3, 6, 9, 16, 22],
037     ...           [10, 13, 14, 17, 24],
038     ...           [18, 21, 23, 26, 30]]
039     >>> target = 1
040     >>> mat_bin_search(target, matrix)
041     [0, 0]
042     >>> target = 34
043     >>> mat_bin_search(target, matrix)
UNEXPECTED EXCEPTION: IndexError('list index out of range')
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.7/x64/lib/python3.10/doctest.py", line 1350, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest matrix.matrix_binary_search.mat_bin_search[4]>", line 1, in <module>
  File "/home/runner/work/Python/Python/matrix/matrix_binary_search.py", line 49, in mat_bin_search
    while matrix[index][0] < value:
IndexError: list index out of range
/home/runner/work/Python/Python/matrix/matrix_binary_search.py:43: UnexpectedException
```

File: matrix/matrix_binary_search.py
Comment (cclauss) [inline matrix/matrix_binary_search.py:51]:
```suggestion
    while matrix[index][0] < value and index <= len(matrix):
```

File: matrix/matrix_binary_search.py
Comment (CaedenPH) [inline matrix/matrix_binary_search.py:13]:
```suggestion
```
You aren't passing `target`, so you don't use the variable

File: matrix/matrix_binary_search.py
Comment (CaedenPH) [inline matrix/matrix_binary_search.py:10]:
```suggestion
```

File: matrix/matrix_binary_search.py
Comment (CaedenPH) [inline matrix/matrix_binary_search.py:1]:
```suggestion
def binary_search(array: list, value: int) -> int:
```
I don't think `lower_bound` and `upper_bound` parameters do much. If the person wants to change the upper and lower bound they should pass the list indexed to the upper and lower bound, such as 

`>>> binary_search([1,2,3,4,5][1:3])`

File: matrix/matrix_binary_search.py
Comment (CaedenPH) [inline matrix/matrix_binary_search.py:5]:
```suggestion
    array: A 1d sorted array
```

**Comment by SUGARXUST** [comment]
Is this still relevant? I would like to work on this.

**Comment by anuragshuklajec** [comment]
Please assign this to me !

**Comment by cclauss** [comment]
Please looks at the `Details` buttons next to the failing tests.
```
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

matrix/matrix_binary_search.py:13:14: F821 undefined name 'lower_bound'
matrix/matrix_binary_search.py:13:28: F821 undefined name 'upper_bound'
matrix/matrix_binary_search.py:16:8: F821 undefined name 'lower_bound'
matrix/matrix_binary_search.py:16:23: F821 undefined name 'upper_bound'
matrix/matrix_binary_search.py:19:44: F821 undefined name 'upper_bound'
matrix/matrix_binary_search.py:21:36: F821 undefined name 'lower_bound'

mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

matrix/matrix_binary_search.py:13: error: Name "lower_bound" is not defined
matrix/matrix_binary_search.py:13: error: Name "upper_bound" is not defined
matrix/matrix_binary_search.py:16: error: Name "lower_bound" is not defined
matrix/matrix_binary_search.py:16: error: Name "upper_bound" is not defined
matrix/matrix_binary_search.py:19: error: Name "upper_bound" is not defined
matrix/matrix_binary_search.py:21: error: Name "lower_bound" is not defined
Found 6 errors in 1 file (checked 1041 source files)
```
FAILED matrix/matrix_binary_search.py::matrix.matrix_binary_search.binary_search
FAILED matrix/matrix_binary_search.py::matrix.matrix_binary_search.mat_bin_search

**Comment by anuragshuklajec** [comment]
Unable to locate the file !
I have fixed errors in my local machine.
Allow me to contribute

**Comment by anuragshuklajec** [comment]
I am ready with the solution kindly help me with the file path

**Comment by cclauss** [comment]
Please read CONTRIBUTING.md.

Hacktoberfest is strenuous for our team of active volunteer maintainers. That small team now has ___200+ open pull requests___ to review.  We have __high standards for our submissions__ which are described in [`CONTRIBUTING.md`](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#contributing-guidelines).  Our continuous integration tests help to ensure that our standards are met before we review and merge deserving pull requests.  This is further complicated by the fact that we already have [lots of algorithms](https://github.com/TheAlgorithms/Python/blob/master/DIRECTORY.md) and would like to avoid duplicates.

For these reasons, I will **close all open pull requests failing tests** so that our unpaid team has a smaller mountain to climb.  If you want us to land your pull request, please read [`CONTRIBUTING.md`](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#contributing-guidelines) _carefully_ and try again.

**Pull request closed without merging**
